lua_modules/mcp23017/mcp23017.lua

```diff
@@ -0,0 +1,300 @@
+--[[
+
+	This Lua module provides access to the MCP23017 module.
+
+    The MCP23017 is a port expander and provides 16 channels for inputs and outputs.
+    Up to 8 devices (128 channels) are possible by the configurable address (A0 - A2 Pins).
+
+    Due to the 16 channels, 2 bytes are required for switching outputs or reading input signals. These are A and B.
+    A single pin can be set or a whole byte. The numbering of the individual pins starts at 0 and ends with 15.
+    The numbers up to 7 are on register GPIO A and from 8 to 15 on GPIO B.
+
+	The module requires `i2c` and `bit` C module built into firmware.
+
+
+	@author Marcel P. | Plomi.net
+	@github https://github.com/plomi-net
+
+	@version 1.0.0
+
+]]
+
+local i2c, string = i2c, string
+local issetBit = bit.isset
+local setBit = bit.set
+local clearBit = bit.clear
+
+local MCP23017_ADDRESS = 0x20 -- (0x20 to 0x27)
+-- MCP23017_SDA = 2 -- D2/GPIO 4
+-- MCP23017_SCL = 1 -- D1/GPIO 5
+
+local OUTPUT = false
+local INPUT = true
+local HIGH = true
+local LOW = false
+
+-- registers (not used registers are commented out)
+local MCP23017_IODIRA = 0x00
+local MCP23017_IODIRB = 0x01
+local MCP23017_DEFVALA = 0x06
+local MCP23017_DEFVALB = 0x07
+local MCP23017_GPIOA = 0x12
+local MCP23017_GPIOB = 0x13
+
+--[[
+local MCP23017_IPOLA = 0x02
+local MCP23017_IPOLB = 0x03
+local MCP23017_GPINTENA = 0x04
+local MCP23017_GPINTENB = 0x05
+local MCP23017_DEFVALA = 0x06
+local MCP23017_DEFVALB = 0x07
+local MCP23017_INTCONA = 0x08
+local MCP23017_INTCONB = 0x09
+local MCP23017_IOCON = 0x0A
+local MCP23017_IOCON2 = 0x0B
+local MCP23017_GPPUA = 0x0C
+local MCP23017_GPPUB = 0x0D
+local MCP23017_INTFA = 0x0E
+local MCP23017_INTFB = 0x0F
+
+local MCP23017_INTCAPA = 0x10
+local MCP23017_INTCAPB = 0x11
+local MCP23017_OLATA = 0x14
+local MCP23017_OLATB = 0x15
+]]
+
+-- internal
+local MCP23017_id = 0
+local MCP23017_SPEED = i2c.SLOW
+local MCP23017_DEVICE_OK = false
```

Comment:
Making these, and `MCP23017_ADDRESS` above, global like this means that, despite appearances (`_G[moduleName or 'mcp23017'] = M`), this module communicates with only one device and attempting to initialize a second can break the first.

Please switch to a "factory" style module, where the result is a function that yields a Lua table with methods, possibly via a metatable.

Reply:
@nwf 
Creating the module in factory style is a very good idea.
Unfortunately I have a little bit of trouble to implement this in Lua. 

Do you have a good example where the objects are not overwritten?
Thanks a lot

Reply:
Apologies for letting this fall off my radar.  An example of a "factory" module is the new LiquidCrystal module: https://github.com/nodemcu/nodemcu-firmware/blob/master/lua_modules/liquidcrystal/liquidcrystal.lua

Reply:
@nwf thanks for your reply. that happens. Good things need a while.
i've updated the module to factory style. Your practice tip helped me a lot.
You are welcome to check if it is proper now.
many thanks.